### PR TITLE
clutil refactor

### PR DESCRIPTION
**Pull request by deanlee**

### CODE REVIEW

**Comment by pd0wm** [comment]
We can just get rid of all the caching code. That should make it a lot simpler.

We used that in the past when modeld wasn't open source and we had our own openCL kernels that we didn't want to ship the source from. But now it just adds complexity and it's probably not worth the very minor startup speedup.

**Comment by deanlee** [comment]
OIC

File: selfdrive/common/clutil.cc [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+  CL_CHECK(clGetPlatformInfo(platform, param_name, 0, NULL, &size));
+  std::string ret;
+  ret.resize(size, '\0');
+  CL_CHECK(clGetPlatformInfo(platform, param_name, size, &ret[0], NULL));
+  return ret;
+}
+
+void cl_print_info(cl_platform_id platform, cl_device_id device) {
+  std::string info = get_platform_info(platform, CL_PLATFORM_VENDOR);
+  printf("vendor: '%s'\n", info.c_str());
```
Comment (pd0wm) [inline selfdrive/common/clutil.cc:24]: `std::cout << "vendor: " << info << std::endl;`?

Comment (pd0wm) [reply selfdrive/common/clutil.cc:24]: Or just skip the temp variable altogether:
`std::cout << "vendor: " << get_platform_info(platform, CL_PLATFORM_VENDOR) << std::endl;

**Comment by pd0wm** [comment]
You can also get rid of the defines in the Sconscript: https://github.com/commaai/openpilot/blob/master/selfdrive/common/SConscript#L33

File: selfdrive/common/clutil.cc [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+  free(src);
+  if (int err = clBuildProgram(prg, 1, &device_id, args, NULL, NULL); err != 0) {
+    cl_print_build_errors(prg, device_id);
+    assert(0);
+  }
+  return prg;
+}
+
+// Given a cl code and return a string represenation
+const char* cl_get_error_string(int err) {
```
Comment (pd0wm) [inline selfdrive/common/clutil.cc:119]: Maybe use a macro like in this example: https://gist.github.com/allanmac/9328bb2d6a99b86883195f8f78fd1b93

Then we don't risk putting the wrong exit code with the wrong string.

File: selfdrive/common/clutil.cc [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    cl_print_info(platform_ids[i], device_id);
+    return device_id;
+  }
+  printf("No valid openCL platform found\n");
+  assert(0);
+  return nullptr;
+}
+
+cl_program cl_program_from_file(cl_context ctx, cl_device_id device_id, const char* path, const char* args) {
+  char* src = (char*)read_file(path, nullptr);
```
Comment (pd0wm) [inline selfdrive/common/clutil.cc:107]: you could use the c++ version from utilpp

File: selfdrive/common/clutil.cc (hunk 1 above)
Comment (deanlee) [reply selfdrive/common/clutil.cc:24]: cool

**Comment by deanlee** [comment]
added struct CLContext to reduce the parameters of the related functions

**Comment by pd0wm** [comment]
Maybe do the CLContext struct in a separate PR? That's probably easier to review on it's own.

**Comment by deanlee** [comment]
the following err codes are undefied
      ```
CL_ERR_TO_STR(CL_INVALID_D3D10_DEVICE_KHR);
      CL_ERR_TO_STR(CL_INVALID_D3D10_RESOURCE_KHR);
      CL_ERR_TO_STR(CL_D3D10_RESOURCE_ALREADY_ACQUIRED_KHR);
      CL_ERR_TO_STR(CL_D3D10_RESOURCE_NOT_ACQUIRED_KHR);
      CL_ERR_TO_STR(CL_INVALID_D3D11_DEVICE_KHR);
      CL_ERR_TO_STR(CL_INVALID_D3D11_RESOURCE_KHR);
      CL_ERR_TO_STR(CL_D3D11_RESOURCE_ALREADY_ACQUIRED_KHR);
      CL_ERR_TO_STR(CL_D3D11_RESOURCE_NOT_ACQUIRED_KHR);
      CL_ERR_TO_STR(CL_INVALID_DX9_MEDIA_ADAPTER_KHR);
      CL_ERR_TO_STR(CL_INVALID_DX9_MEDIA_SURFACE_KHR);
      CL_ERR_TO_STR(CL_DX9_MEDIA_SURFACE_ALREADY_ACQUIRED_KHR);
      CL_ERR_TO_STR(CL_DX9_MEDIA_SURFACE_NOT_ACQUIRED_KHR);
      CL_ERR_TO_STR(CL_INVALID_EGL_OBJECT_KHR);
      CL_ERR_TO_STR(CL_EGL_RESOURCE_NOT_ACQUIRED_KHR);
      CL_ERR_TO_STR(CL_PLATFORM_NOT_FOUND_KHR);
      CL_ERR_TO_STR(CL_DEVICE_PARTITION_FAILED_EXT);
      CL_ERR_TO_STR(CL_INVALID_PARTITION_COUNT_EXT);
      CL_ERR_TO_STR(CL_INVALID_PARTITION_NAME_EXT);
      CL_ERR_TO_STR(CL_INVALID_ACCELERATOR_INTEL);
      CL_ERR_TO_STR(CL_INVALID_ACCELERATOR_TYPE_INTEL);
      CL_ERR_TO_STR(CL_INVALID_ACCELERATOR_DESCRIPTOR_INTEL);
      CL_ERR_TO_STR(CL_ACCELERATOR_TYPE_NOT_SUPPORTED_INTEL);
      CL_ERR_TO_STR(CL_INVALID_GL_SHAREGROUP_REFERENCE_KHR);
      CL_ERR_TO_STR(CL_INVALID_VA_API_MEDIA_ADAPTER_INTEL);
      CL_ERR_TO_STR(CL_INVALID_VA_API_MEDIA_SURFACE_INTEL);
      CL_ERR_TO_STR(CL_VA_API_MEDIA_SURFACE_ALREADY_ACQUIRED_INTEL);
      CL_ERR_TO_STR(CL_VA_API_MEDIA_SURFACE_NOT_ACQUIRED_INTEL);
```

**Comment by deanlee** [comment]
too many undefined error codes.I think it's better not to use CL_ERR_TO_STR this time.

**Comment by pd0wm** [comment]
The define is just a helper function. Can't you mix them with the manual cases?

File: selfdrive/common/clutil.cc [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  }
+  std::cout << "No valid openCL platform found" << std::endl;
+  assert(0);
+  return nullptr;
+}
+
+cl_program cl_program_from_file(cl_context ctx, cl_device_id device_id, const char* path, const char* args) {
+  std::string src = util::read_file(path);
+  assert(src.length() > 0);
+  cl_program prg = CL_CHECK_ERR(clCreateProgramWithSource(ctx, 1, (const char*[]){&src[0]}, NULL, &err));
```
Comment (pd0wm) [inline selfdrive/common/clutil.cc:82]: I don't think `&src[0]` is guaranteed to be null terminated. Use `src.c_str()`

File: selfdrive/common/clutil.cc [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+#include "utilpp.h"
+
+namespace {  // helper functions
+
+template <typename Func, typename Id, typename Name>
+std::string get_info(Func get_info_func, Id id, Name param_name) {
+  size_t size = 0;
+  CL_CHECK(get_info_func(id, param_name, 0, NULL, &size));
+  std::string info(size, '\0');
+  CL_CHECK(get_info_func(id, param_name, size, &info[0], NULL));
```
Comment (pd0wm) [inline selfdrive/common/clutil.cc:19]: Probably cleaner to use `info.data`.

File: selfdrive/common/clutil.cc [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+  }
+
+  std::cout << "vendor: " << get_platform_info(platform, CL_PLATFORM_VENDOR) << std::endl
+            << "platform version: " << get_platform_info(platform, CL_PLATFORM_VERSION) << std::endl
+            << "profile: " << get_platform_info(platform, CL_PLATFORM_PROFILE) << std::endl
+            << "extensions: " << get_platform_info(platform, CL_PLATFORM_EXTENSIONS) << std::endl
+            << "name :" << get_device_info(device, CL_DEVICE_NAME) << std::endl
+            << "device version :" << get_device_info(device, CL_DEVICE_VERSION) << std::endl
+            << "max work group size :" << work_group_size << std::endl
+            << "type = " << std::hex << std::showbase << device_type << " = " << type_str << std::endl;
```
Comment (pd0wm) [inline selfdrive/common/clutil.cc:44]: You're not restoring the output properties of `std::cout`. Won't everything after this be in hex? Printing in hex is not that important, so maybe just leave it as is?

**Pull request merged**